File: macos/Tests/VPSQuotaCoreTests/CrossPlatformConfigTests.swift
import Foundation
import Testing
@testable import VPSQuotaCore

/// config.json 是 macOS 与 Windows 两端共用的格式，用户可以在两台机器之间直接拷贝。
/// 这组测试把字段名和枚举取值钉死 —— 任何一端改了命名，这里都会先红。
@Suite("跨平台配置格式")
struct CrossPlatformConfigTests {

    private func encode(_ config: AppConfig) throws -> [String: Any] {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        let data = try encoder.encode(config)
        return try #require(
            JSONSerialization.jsonObject(with: data) as? [String: Any]
        )
    }

    @Test("枚举一律编码成小写字符串，与 C# 端的 camelCase 策略一致")
    func enumsEncodeAsLowercaseStrings() throws {
        let config = AppConfig(servers: [
            ServerConfig(id: "a", name: "东京", provider: .vultr,
                         meterMode: .outbound, unitBase: .binary),
            ServerConfig(id: "b", name: "洛杉矶", provider: .ssh,
                         meterMode: .sum, unitBase: .decimal),
        ])
        let json = try encode(config)
        let servers = try #require(json["servers"] as? [[String: Any]])

        #expect(servers[0]["provider"] as? String == "vultr")
        #expect(servers[0]["meterMode"] as? String == "outbound")
        #expect(servers[0]["unitBase"] as? String == "binary")
        #expect(servers[1]["provider"] as? String == "ssh")
        #expect(servers[1]["meterMode"] as? String == "sum")
        #expect(servers[1]["unitBase"] as? String == "decimal")
    }

    @Test("顶层与服务器的字段名与 C# 端逐字对应")
    func fieldNamesMatchWindows() throws {
        let config = AppConfig(refreshIntervalMinutes: 60, servers: [
            ServerConfig(id: "b", name: "洛杉矶", provider: .ssh,
                         quotaGB: 1000, meterMode: .sum, resetDay: 15,
                         unitBase: .binary,
                         sshHost: "1.2.3.4", sshPort: 22, sshUser: "root",
                         sshKeyPath: "~/.ssh/id_ed25519", interface: "eth0"),
        ])
        let json = try encode(config)
        #expect(Set(json.keys) == ["refreshIntervalMinutes", "servers"])

        let server = try #require((json["servers"] as? [[String: Any]])?.first)
        // C# 端的 JsonNamingPolicy.CamelCase 只把首字母小写，
        // 所以 QuotaGB → quotaGB（而不是 quotaGb），这一条最容易写错。
        #expect(Set(server.keys) == [
            "id", "name", "provider", "quotaGB", "meterMode", "resetDay", "unitBase",
            "sshHost", "sshPort", "sshUser", "sshKeyPath", "interface",
        ])
    }

    @Test("能读入 Windows 端写出的配置")
    func decodesWindowsWrittenConfig() throws {
        // 这份 JSON 是按 C# 端 JsonNamingPolicy.CamelCase + 枚举转换器的输出格式手写的
        let json = """
        {
          "refreshIntervalMinutes": 360,
          "servers": [
            {
              "id": "vultr-tokyo",
              "name": "Vultr 东京",
              "provider": "vultr",
              "quotaGB": 0,
              "meterMode": "outbound",
              "resetDay": 1,
              "unitBase": "binary",
              "vultrInstanceId": "cb676a46-66fd-4dfb-b839-443f2e6c0b60"
            },
            {
              "id": "dmit-lax",
              "name": "DMIT 洛杉矶",
              "provider": "ssh",
              "quotaGB": 1000,
              "meterMode": "sum",
              "resetDay": 15,
              "unitBase": "decimal",
              "sshHost": "1.2.3.4",
              "sshPort": 2222,
              "sshUser": "root",
              "sshKeyPath": "~/.ssh/id_ed25519",
              "interface": "eth0"
            }
          ]
        }
        """
        let config = try JSONDecoder().decode(AppConfig.self, from: Data(json.utf8))

        #expect(config.refreshIntervalMinutes == 360)
        #expect(config.servers.count == 2)
        #expect(config.servers[0].provider == .vultr)
        #expect(config.servers[0].meterMode == .outbound)
        #expect(config.servers[0].unitBase == .binary)
        #expect(config.servers[0].quotaGB == 0)
        #expect(config.servers[1].provider == .ssh)
        #expect(config.servers[1].meterMode == .sum)
        #expect(config.servers[1].unitBase == .decimal)
        #expect(config.servers[1].sshPort == 2222)
        #expect(config.servers[1].interface == "eth0")
    }

    @Test("示例配置文件本身能被解析")
    func exampleConfigIsValid() throws {
        let url = URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent()   // VPSQuotaCoreTests
            .deletingLastPathComponent()   // Tests
            .deletingLastPathComponent()   // macos
            .deletingLastPathComponent()   // 仓库根目录
            .appendingPathComponent("shared/config.example.json")
        let data = try Data(contentsOf: url)
        let config = try JSONDecoder().decode(AppConfig.self, from: data)
        #expect(config.servers.count == 4)
        #expect(config.refreshIntervalMinutes == 60)
    }
}

File: windows/VpsQuota.Tests/CrossPlatformConfigTests.cs
namespace VpsQuota.Tests;

using System.IO;
using System.Runtime.CompilerServices;
using System.Text.Json;
using VpsQuota.Core;
using VpsQuota.Models;
using VpsQuota.Storage;
using Xunit;

/// <summary>
/// 跨端配置格式（需求 G2、验收标准第 2 条：config.json 拷过去要能直接用）。
/// </summary>
/// <remarks>
/// 与 macOS 端的 <c>CrossPlatformConfigTests.swift</c> 对应。两边都读同一份
/// <c>shared/config.example.json</c>，任何一端改了字段名或默认值，对面就会红。
/// </remarks>
public class CrossPlatformConfigTests
{
    private static string RepoRoot([CallerFilePath] string? thisFile = null)
    {
        var directory = Path.GetDirectoryName(thisFile)!;
        return Path.GetFullPath(Path.Combine(directory, "..", ".."));
    }

    private static AppConfig LoadExample() =>
        new ConfigStore(Path.Combine(RepoRoot(), "shared", "config.example.json")).Load();

    [Fact(DisplayName = "示例配置文件本身能被解析")]
    public void ExampleParses()
    {
        var config = LoadExample();

        Assert.Equal(60, config.RefreshIntervalMinutes);
        Assert.Equal("dmit-lax", config.MenuBarServerId);
        Assert.True(config.MenuBarShowsRemaining);
        Assert.Equal(2, config.Servers.Count);
    }

    [Fact(DisplayName = "枚举一律解析自小写字符串，与 macOS 端的 RawValue 一致")]
    public void EnumsUseCamelCase()
    {
        var config = LoadExample();

        var vultr = config.Servers.Single(s => s.Id == "vultr-singapore");
        Assert.Equal(ProviderKind.Vultr, vultr.Provider);
        Assert.Equal(MeterMode.Outbound, vultr.MeterMode);
        Assert.Equal(UnitBase.Binary, vultr.UnitBase);

        var dmit = config.Servers.Single(s => s.Id == "dmit-lax");
        Assert.Equal(ProviderKind.Ssh, dmit.Provider);
        Assert.Equal(MeterMode.Sum, dmit.MeterMode);
        Assert.Equal(10, dmit.ResetDay);
        Assert.Equal("eth0", dmit.Interface);
        Assert.Equal(22, dmit.SshPort);
    }

    [Fact(DisplayName = "能读入带起始已用量基准的配置")]
    public void ReadsUsageBaseline()
    {
        var dmit = LoadExample().Servers.Single(s => s.Id == "dmit-lax");

        Assert.NotNull(dmit.UsageBaseline);
        Assert.Equal("2026-09-10", dmit.UsageBaseline!.PeriodStart);
        Assert.Equal(320.5, dmit.UsageBaseline.UsedGB);
    }

    [Fact(DisplayName = "写出的 JSON 用驼峰字段名，且不含 null")]
    public void WritesCamelCaseWithoutNulls()
    {
        var path = Path.Combine(Path.GetTempPath(), "vpsquota-cfg-" + Guid.NewGuid().ToString("n") + ".json");
        var store = new ConfigStore(path);

        store.Save(new AppConfig
        {
            RefreshIntervalMinutes = 60,
            Servers = new List<ServerConfig>
            {
                new()
                {
                    Id = "s1", Name = "测试", Provider = ProviderKind.Ssh,
                    QuotaGB = 1000, MeterMode = MeterMode.Sum, ResetDay = 10,
                },
            },
        });

        var json = File.ReadAllText(path);
        try { File.Delete(path); } catch (IOException) { }

        Assert.Contains("\"refreshIntervalMinutes\"", json);
        Assert.Contains("\"quotaGB\"", json);
        Assert.Contains("\"meterMode\": \"sum\"", json);
        Assert.Contains("\"provider\": \"ssh\"", json);

        // Swift 的 JSONEncoder 省略 nil。这边若写出 null，
        // 不但两端产物不是同一份文本，自己下次读还会因为非 nullable 值类型而抛异常。
        Assert.DoesNotContain("null", json);
        // menuBarServerId 没指定，不该出现在文件里
        Assert.DoesNotContain("menuBarServerId", json);
    }

    [Fact(DisplayName = "缺必填字段的条目被跳过，其余服务器仍然可用")]
    public void SkipsBrokenEntries()
    {
        var json = """
        {
          "refreshIntervalMinutes": 60,
          "servers": [
            { "id": "ok-1", "name": "正常", "provider": "ssh", "quotaGB": 1000 },
            { "name": "缺 id 和 provider", "quotaGB": 500 },
            { "id": "ok-2", "name": "也正常", "provider": "vultr" }
          ]
        }
        """;
        var path = Path.Combine(Path.GetTempPath(), "vpsquota-broken-" + Guid.NewGuid().ToString("n") + ".json");
        File.WriteAllText(path, json);

        var store = new ConfigStore(path);
        var config = store.Load();
        try { File.Delete(path); } catch (IOException) { }

        Assert.Equal(new[] { "ok-1", "ok-2" }, config.Servers.Select(s => s.Id));
        Assert.Equal(1, store.SkippedServerCount);
    }

    [Fact(DisplayName = "显式 null 的字段退回默认值")]
    public void ExplicitNullFallsBackToDefault()
    {
        var json = """
        {
          "refreshIntervalMinutes": 60,
          "servers": [
            {
              "id": "s1", "name": "测试", "provider": "ssh",
              "quotaGB": null, "resetDay": null, "unitBase": null,
              "sshPort": null, "usageBaseline": null
            }
          ]
        }
        """;
        var path = Path.Combine(Path.GetTempPath(), "vpsquota-null-" + Guid.NewGuid().ToString("n") + ".json");
        File.WriteAllText(path, json);

        var config = new ConfigStore(path).Load();
        try { File.Delete(path); } catch (IOException) { }

        var server = Assert.Single(config.Servers);
        Assert.Equal(0, server.QuotaGB);
        Assert.Equal(1, server.ResetDay);
        Assert.Equal(UnitBase.Binary, server.UnitBase);
        Assert.Null(server.SshPort);
        Assert.Null(server.UsageBaseline);
    }

    [Fact(DisplayName = "字段名区分大小写，与 macOS 端一样严格")]
    public void PropertyNamesAreCaseSensitive()
    {
        // {"Provider": …} 这种写法以前在 Windows 上能过、拷到 macOS 才炸。
        // 现在两端一样严格：provider 认不出来 → 视为缺必填字段 → 跳过该条目。
        var json = """
        { "servers": [ { "id": "s1", "name": "测试", "Provider": "ssh" } ] }
        """;
        var path = Path.Combine(Path.GetTempPath(), "vpsquota-case-" + Guid.NewGuid().ToString("n") + ".json");
        File.WriteAllText(path, json);

        var store = new ConfigStore(path);
        var config = store.Load();
        try { File.Delete(path); } catch (IOException) { }

        Assert.Empty(config.Servers);
        Assert.Equal(1, store.SkippedServerCount);
    }
}
